plugins/preview/plugin.js

```diff
@@ -8,128 +8,262 @@
-						'<title>' + editor.lang.preview.preview + '</title>' +
-						CKEDITOR.tools.buildStyleHtml( editor.config.contentsCss ) +
-					'</head>' + bodyHtml +
-						editor.getData() +
-					'</body></html>';
+	var pluginName = 'preview',
+		previewCmd;
+
+	// Register a plugin named "preview".
+	CKEDITOR.plugins.add( pluginName, {
+		// jscs:disable maximumLineLength
+		lang: 'af,ar,az,bg,bn,bs,ca,cs,cy,da,de,de-ch,el,en,en-au,en-ca,en-gb,eo,es,es-mx,et,eu,fa,fi,fo,fr,fr-ca,gl,gu,he,hi,hr,hu,id,is,it,ja,ka,km,ko,ku,lt,lv,mk,mn,ms,nb,nl,no,oc,pl,pt,pt-br,ro,ru,si,sk,sl,sq,sr,sr-latn,sv,th,tr,tt,ug,uk,vi,zh,zh-cn', // %REMOVE_LINE_CORE%
+		// jscs:enable maximumLineLength
+		icons: 'preview,preview-rtl', // %REMOVE_LINE_CORE%
+		hidpi: true, // %REMOVE_LINE_CORE%
+		init: function( editor ) {
+			var pluginPath = CKEDITOR.plugins.getPath( 'preview' );
+
+			editor.addCommand( pluginName, previewCmd );
+			editor.ui.addButton && editor.ui.addButton( 'Preview', {
+				label: editor.lang.preview.preview,
+				command: pluginName,
+				toolbar: 'document,40'
+			} );
+
+			if ( !CKEDITOR.env.webkit || CKEDITOR.env.chrome ) {
+				return;
 			}
 
-			var iWidth = 640,
-				// 800 * 0.8,
-				iHeight = 420,
-				// 600 * 0.7,
-				iLeft = 80; // (800 - 0.8 * 800) /2 = 800 * 0.1.
-			try {
-				var screen = window.screen;
-				iWidth = Math.round( screen.width * 0.8 );
-				iHeight = Math.round( screen.height * 0.7 );
-				iLeft = Math.round( screen.width * 0.1 );
-			} catch ( e ) {}
+			// If Safari does not have print styles in cache, it won't
+			// render the printed content. Therefore we need to preload the styles.
+			preloadStyles( editor.config.contentsCss );
+			preloadStyles( pluginPath + 'styles/screen.css' );
+		}
+	} );
+
+	/**
+	 * Namespace providing a set of helper functions for working with print preview, exposed by the
+	 * [Preview](https://ckeditor.com/cke4/addon/tableselection) plugin.
+	 *
+	 * @since 4.14.0
+	 * @singleton
+	 * @class CKEDITOR.plugins.preview
+	 */
+	CKEDITOR.plugins.preview = {
+		/**
+		 * Generates print preview for the given editor.
+		 *
+		 * @param {CKEDITOR.editor} editor Editor's instance.
+		 * @returns {CKEDITOR.dom.window} Newly created window, which contains the preview.
+		 */
+		createPreview: function( editor ) {
+			var previewHtml = createPreviewHtml( editor ),
+				eventData = { dataValue: previewHtml },
+				windowDimensions = getWindowDimensions(),
+				// For IE we should use window.location rather than setting url in window.open (https://dev.ckeditor.com/ticket/11146).
+				previewLocation = getPreviewLocation(),
+				previewUrl = getPreviewUrl(),
+				previewWindow,
+				doc;
 
 			// (https://dev.ckeditor.com/ticket/9907) Allow data manipulation before preview is displayed.
 			// Also don't open the preview window when event cancelled.
-			if ( editor.fire( 'contentPreview', eventData = { dataValue: sHTML } ) === false )
+			if ( editor.fire( 'contentPreview', eventData ) === false ) {
```

Comment:
I'm wondering how this case can be reflected in the print plugin. Previously, the print plugin didn't depend on the preview plugin and has been printing current editor HTML. So, if someone highly customizes `preview` with `contentPreview` event, he will get totally different HTML after these changes. Let's say, someone has been adding custom headers to the preview HTML but assumes that print plugin is a separate feature that shouldn't depend on this event. I'm wondering if we could exclude additional, `contentPrint` event for that case.

On the other hand, `contentPreview` is recommended to use to prevent XSS (https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR.html#event-contentPreview), so if we stop executing this function, it may lead to security issues with the print command which is even the worse scenario.

Maybe, at least, we could pass some information if it's a preview or print feature by passing some flag in `contentPreview` event, so someone can decide what to do? Could be done by listening on `contentPreview` with high priority in print plugin to avoid dome dependency issues, and adding some additional flag or enum. WDYT?

Reply:
I don't understand what problem does the new event fix? It would make the whole implementation much more convoluted and lead to producing two versions of content – one for the preview and one for printing. So we would be basically back to the starting point.

Reply:
> Maybe, at least, we could pass some information if it's a preview or print feature by passing some flag in contentPreview event, so someone can decide what to do? Could be done by listening on contentPreview with high priority in print plugin to avoid dependency issues, and adding some additional flag or enum. WDYT?

Someone could add some fancy HTML to the preview page (like a logo) which shouldn't be visible on the print page. Additional information in the event object could slightly minimalize breaking changes in that case, so it's still possible to customize preview content but not print (and vice versa).

Reply:
Ok, so as we already talked F2F it's rather an edge case as `contentPreview` event should be mostly used to sanitize HTML which seems also important when using a print plugin. Let's see if someone directly requests such improvement instead of pre-optimizing.